service/history/shard/context.go

```diff
@@ -1617,3 +1621,74 @@ func acquireShard(
 
 	return context, nil
 }
+
+func (s *contextImpl) getEventsFromWorkflowSnapshot(snapshot *persistence.WorkflowSnapshot) []simulation.E {
+	if snapshot == nil {
+		return nil
+	}
+	var events []simulation.E
+	for category, tasks := range snapshot.TasksByCategory {
+		for _, task := range tasks {
+			events = append(events, simulation.E{
+				EventName:  simulation.EventNameCreateHistoryTask,
+				Host:       s.config.HostName,
+				ShardID:    s.shardID,
+				DomainID:   task.GetDomainID(),
+				WorkflowID: task.GetWorkflowID(),
+				RunID:      task.GetRunID(),
+				Payload: map[string]any{
+					"task_category": category.Name(),
+					"task_type":     task.GetTaskType(),
+					"task_key":      task.GetTaskKey(),
+				},
+			})
+		}
+	}
+	return events
+}
+
+func (s *contextImpl) getEventsFromWorkflowMutation(mutation *persistence.WorkflowMutation) []simulation.E {
+	if mutation == nil {
+		return nil
+	}
+	var events []simulation.E
+	for category, tasks := range mutation.TasksByCategory {
+		for _, task := range tasks {
+			events = append(events, simulation.E{
+				EventName:  simulation.EventNameCreateHistoryTask,
+				Host:       s.config.HostName,
+				ShardID:    s.shardID,
+				DomainID:   task.GetDomainID(),
+				WorkflowID: task.GetWorkflowID(),
+				RunID:      task.GetRunID(),
+				Payload: map[string]any{
+					"task_category": category.Name(),
+					"task_type":     task.GetTaskType(),
+					"task_key":      task.GetTaskKey(),
+				},
+			})
+		}
+	}
+	return events
+}
+
+func (s *contextImpl) logCreateWorkflowExecutionEvents(request *persistence.CreateWorkflowExecutionRequest) {
+	events := s.getEventsFromWorkflowSnapshot(&request.NewWorkflowSnapshot)
+	simulation.LogEvents(events...)
+}
+
+func (s *contextImpl) logUpdateWorkflowExecutionEvents(request *persistence.UpdateWorkflowExecutionRequest) {
+	events := s.getEventsFromWorkflowMutation(&request.UpdateWorkflowMutation)
+	simulation.LogEvents(events...)
+	events = s.getEventsFromWorkflowSnapshot(request.NewWorkflowSnapshot)
+	simulation.LogEvents(events...)
```

Comment:
There's quite a bit of overhead to prepare these event structures. Consider exposing `simulation.Enabled() bool` to skip all this